util/ribbon_impl.h

```diff
@@ -49,10 +53,22 @@ namespace ribbon {
 //   // construction.
 //   static constexpr bool kIsFilter;
 //
+//   // When true, enables a special "homogeneous" filter implementation that
+//   // is slightly faster to construct, and never fails to construct though
+//   // FP rate can quickly explode in cases where corresponding
+//   // non-homogeneous filter would fail (or nearly fail?) to construct.
+//   // For smaller filters, you can configure with ConstructionFailureChance
+//   // smaller than desired FP rate to largely counteract this effect.
+//   // TODO: configuring Homogeneous Ribbon for arbitrarily large filters
+//   // based on data from OptimizeHomogAtScale
+//   static constexpr bool kHomogeneous;
```

Comment:
Why not just have a special hash function always return 0?

Reply:
I don't know exactly what you mean, but for sharing the same general algorithms, GetResultRowFromHash always returns 0 when kHomogeneous. Regardless of kHomogeneous, we work from a 64-bit hash of the input key for start position and coefficient row block, being smart about how we re-use/re-mix that hash info.

Reply:
This flag also controls whether we populate free variable rows in the solution pseudorandomly or simply as zeros.

I would agree with this: in RocksDB, we generally prefer dynamic if-based configuration to static template-based configuration. This is good for simplicity, easier testing, avoiding explosion of compiled code size, etc. We do not optimize performance at all software engineering costs.

If this were a case of implementing a known algorithm for use in RocksDB, where we would just implement the subset of features we want, I would generally prefer that approach, like I did with the FastLocalBloom. But this is a new algorithm and I wanted to create a generic reference implementation. We aren't yet sure which subset of features we will want to keep for long-term use in RocksDB. And even if we aren't sensitive to performance down to the ns, some people are, and we want to show the best numbers we reasonably can for a paper.